Ingest dev guide (#1415)
stellar/stellar-docs: docs/data/ingest-library/developer_guide/architecture.mdx

---
title: Architecture
sidebar_position: 0
---

> ![Insert architecture diagram]

Reviewer: remove this placeholder on arch diag for now? since https://github.com/stellar/go/issues/5595 will likely change it..

Author: Thanks @sreuland. Could you share the Excalidraw link to this diagram?

Reviewer: oops, I didn't export the full scene for the board, here's the session link - https://excalidraw.com/#room=66624d46471ff1b43439,7OvkGakFf3IC7m9S-2-Cow

Author: Added these 2 diagrams:
 
https://developers-pr1415.previews.kube001.services.stellar-ops.com/docs/data/indexers/build-your-own/ingest-sdk/developer_guide/architecture

https://developers-pr1415.previews.kube001.services.stellar-ops.com/docs/data/indexers/build-your-own/ingest-sdk/developer_guide/ledgerbackends/bufferedstoragebackend

## Package Structure

> The ingest library is broadly split into three main packages:

Reviewer: Should we include mention of HistoryArchive usage via the `github.com/stellar/go/historyarchive` package at this level it's significant for ledger entry aspect. And the xdr bindings from `github.com/stellar/go/xdr` as it contains all data models.

I think it might be worthwhile to include the formal package names for each item mentioned either expressed as code formatted markup like`github.com/stellar/go/ingest/ledgerbackend` or as link to gh repo path [github.com/stellar/go/ingest/ledgerbackend](https://github.com/stellar/go/tree/master/ingest/ledgerbackend)

Author: >I think it might be worthwhile to include the formal package names for each item mentioned either expressed as code formatted markup likegithub.com/stellar/go/ingest/ledgerbackend or as link to gh repo path [github.com/stellar/go/ingest/ledgerbackend](https://github.com/stellar/go/tree/master/ingest/ledgerbackend)

done.

Author: >Should we include mention of HistoryArchive usage via the github.com/stellar/go/historyarchive package at this level it's significant for ledger entry aspect. And the xdr bindings from github.com/stellar/go/xdr as it contains all data models.

I'm not sure about including historyarchives here since it's outside the ingest package at least for now. For xdr, I'll add more details about the data available in ledgerclosemeta object (see Tamir's comment [here](https://github.com/stellar/stellar-docs/pull/1415#discussion_r1992947277)) and will include the link to the xdr package there.

Reviewer: Yes, `historyarchive` is not under `github.com/stellar/go/ingest`, but its functionality is a subset of ingestion development, we reference consuming history archives to get ledger entries as a use case for ingest sdk in the [introduction](https://github.com/stellar/stellar-docs/pull/1422/files#diff-d51c933b9067a53bdc98bad2e9fce260269dd7dad23aa61cb2de1725e16582daR18)

> 1. **Ledger Backends**: A wrapper around sources of ledger data to fetch ledgers from the Stellar network.

Reviewer 2: I'm not sure if wrapper is the right term to use. I would suggest rephrasing the description to something like
```suggestion
1. **Ledger Backends**: Used to stream ledgers from the Stellar network.
```

> 2. **Ledger Readers**: Utility classes that help iterate through individual changes or transactions in a ledger.

Reviewer 2: since the library is written in Go we should avoid the term class because although Go supports object oriented programming it doesn't refer to objects as classes.

```suggestion
2. **Ledger Readers**: Iterators to extract individual changes or transactions from a ledger.
```

> 3. **Ledger Processors**: Functions to process individual changes or transactions and extract meaningful data such as offers, claimable balance, etc.

Reviewer 2: you can mention that the ledger processors are used in conjunction with the ledger readers

Author: done.
